_posts/2019-02-01-aliasing-fields-on-ecto-associations-in-absinthe-for-graphql.md

```diff
@@ -0,0 +1,131 @@
+
+One technique for handling this is to continue presenting the original field, but using data from another source. I would refer to this as “aliasing” or mapping one field’s data into another. The fundamental idea in Absinthe is that the name of the field presented in the API is not required to be the key for the source of data we use.
+
+I’ll cover two scenarios of database refactorings I went through recently, while keeping the API backward compatible:
+
+1. Renaming a column
+2. Moving a column into another table
+
+## Renaming a Column ##
+
+This is a simple example to get us warmed up, but we’ll use this technique in combination with another in our next example.
+
+We’ll start with a table and a column (example from Postgres):
+
+<pre><code class="language-sql">
+CREATE TABLE firmware_versions (
+  id SERIAL,
+  semantic_version_number text
+);
+
+CREATE TABLE devices (
+  id SERIAL,
+  firmware_version_id integer REFERENCES firmware_versions(id),
+  desired_firmware_version_id integer REFERENCES firmware_versions(id)
+);
+</code></pre>
+
+And the Absinthe object type would look something like (I’ve omitted data loaders and resolvers):
+
+<pre><code class="language-elixir">
+@desc "Firmware version"
+object :firmware_version do
+  @desc "Database ID of the firmware version"
+  field(:id, :integer)
+
+  @desc "Firmware version expressed as a semantic version number"
+  field(:semantic_version_number, :string)
+end
+
+@desc "A network-connected device"
+object :device do
+  …
+  @desc "Current firmware version of the device"
+  field(:firmware_version, :firmware_version, resolve: dataloader(Device))
+
+  @desc “Firmware version we are attempting to update to"
+  field(:desired_firmware_version, :firmware_version, resolve: dataloader(Device))
+end
+</code></pre>
+
+### Changing the object type for the new column name ###
+
+What happens when we change the column name from `desired_firmware_version_id` to `requested_firmware_version_id`? An admittedly contrived example, meant to highlight a point. Aside from the normal database migration to alter the table, we need only make a straightforward change in our GraphQL type for the device to maintain the interface for our users.
+
+<pre><code class="language-elixir">
+field(:requested_firmware_version, :firmware_version, resolve: dataloader(Device), name: "desired_firmware_version")
+</code></pre>
+
+Two changes are required. Change the first argument in `field` to match the new name of the column (in this case, an association). And then, add the option for `name` and pass in a string (binary) to match the previous column’s name in the API.
+
+We may also wish to make use of the `deprecate` option available to `field` and pass it a string with the reason for the deprecation.
```

Comment:
Also, we may wish to use the `deprecate` …

Reply:
If you are planning to deprecate the old field name over time, use the `deprecate` option that Absinthe provides.